Add DgElement collection component and CreateElementCollection action

nilsdeppe opened the pull request:
## Proposed changes

This is the last bit of code for GTS nodegroup with no functions of time. I'll submit another PR soon to add support for functions of time. I still need to finish support for LTS.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Add CreateElementCollection action

nilsdeppe committed:
Add DgElementCollection

knelli2 commented on src/Parallel/ArrayCollection/CreateElementCollection.hpp:
```diff
@@ -0,0 +1,201 @@
+/*!
+ * \brief Creates the `DgElementArrayMember`s on the (node)group component.
+ *
+ * First the distribution of elements is computed using
+ * `Parallel::create_elements_using_distribution()`, they they are inserted on
```
`they they` -> `they`

knelli2 commented on src/Parallel/ArrayCollection/CreateElementCollection.hpp:
```diff
@@ -0,0 +1,201 @@
+    }();
+    std::vector<std::pair<ElementId<Dim>, size_t>> my_elements_and_cores{};
+    my_elements_and_cores.reserve(
+        static_cast<size_t>(std::round(static_cast<double>(total_num_elements) /
+                                       static_cast<double>(number_of_nodes))));
```
Could this be just `total_num_elements / number_of_nodes + 1`? It's just a reservation so it's ok if it's a bit bigger.

nilsdeppe replied:
Sure :shrug:

knelli2 commented on src/Parallel/ArrayCollection/CreateElementCollection.hpp:
```diff
@@ -0,0 +1,201 @@
+          }
+        },
+        element_weight, blocks, initial_extents, initial_refinement_levels,
+        quadrature,
+
```
Remove blank line

nilsdeppe replied:
I personally find the long functions easier to parse when there's a logical split between arguments. In this case everything below is related to parallelization/hardware and the above to the domain.

knelli2 replied:
Could you add a comment then instead of a blank line? We do this in some other places and I think it's clearer (a blank line looks like a formatter error to me)

knelli2 reviewed (changes requested):
Also some clang-tidy stuff about structured bindings in lambdas.

@kidder Do you want to look at this?

nilsdeppe commented:
I've rebased and squashed in the couple changes. I don't know what to do about the clang-tidy errors because I think this is an issue with clang-tidy not compiling the code correctly. I can build the code fine and what clang-tidy is reporting is a compilation error. With C++17 or 20 structured bindings are variables that can be captured in lambdas, whereas prior to that there were local names and thus could not be captured. I think the issue has something to do with how clang-tidy is being used. It's either that we are using an older version or that the standard flag is not being handled correctly in the header file context.

knelli2 reviewed (changes requested):
According to the comments of the accepted answer [here](https://stackoverflow.com/questions/46114214/lambda-implicit-capture-fails-with-variable-declared-from-structured-binding), even though the feature is supported in c++17/20, apparently clang (through major version 15) still didn't support this. And I just checked, in the container we use a `clang-tidy` version of 14, so that'd explain why we are getting the errors. Looks like we should update the clang-tidy version we use, but that's beyond this PR.

kidder commented on src/Parallel/ArrayCollection/CreateElementCollection.hpp:
```diff
@@ -0,0 +1,199 @@
+ * - Modifies:
+ *   - `Parallel::Tags::ElementCollection`
+ *   - `Parallel::Tags::ElementLocations<Dim>`
+ */
```
modifies NumberOfElementsTerminated as well

kidder commented on src/Parallel/ArrayCollection/DgElementCollection.hpp:
```diff
@@ -0,0 +1,117 @@
+  auto downcast(const Parallel::DgElementArrayMemberBase<Dim>& base)
+      -> const dg_element_array_member& {
+#ifdef SPECTRE_DEBUG
+    return dynamic_cast<const dg_element_array_member&>(base);
```
what is the reason for doing the cast in this manner?

nilsdeppe replied:
I haven't needed these (thought I would) so just removing them.

nilsdeppe commented:
I've rebased and squashed in fixes. Let me know if there's anything else!

knelli2 commented:
Ignoring macOS timeout and outdated clang-tidy errors
